sdk/src/asset_handlers/riff_io.rs

```diff
@@ -314,38 +526,119 @@ impl RemoteRefEmbed for RiffIO {
+
+        self.embed_reference_to_stream(&mut input_stream, &mut output_stream, embed_ref)
+    }
+
+    fn embed_reference_to_stream(
+        &self,
+        input_stream: &mut dyn CAIRead,
+        output_stream: &mut dyn CAIReadWrite,
+        embed_ref: RemoteRefEmbedType,
     ) -> Result<()> {
         match embed_ref {
             crate::asset_io::RemoteRefEmbedType::Xmp(manifest_uri) => {
-                #[cfg(feature = "xmp_write")]
-                {
-                    match self.riff_format.as_ref() {
-                        "avi" | "wav" => {
-                            crate::embedded_xmp::add_manifest_uri_to_file(asset_path, &manifest_uri)
+                match self.riff_format.as_ref() {
+                    "avi" | "wav" | "webp" => {
+                        if let Some(curr_xmp) = self.read_xmp(input_stream) {
+                            let mut new_xmp = add_provenance(&curr_xmp, &manifest_uri)?;
+                            if new_xmp.len() % 2 == 1 {
+                                // pad if needed to even length
+                                new_xmp.push(' ');
+                            }
+
+                            let top_level_chunks = {
+                                let mut reader = CAIReadWrapper {
+                                    reader: input_stream,
+                                };
+                                Chunk::read(&mut reader, 0)?
+                            };
+
+                            if top_level_chunks.id() != RIFF_ID {
+                                return Err(Error::InvalidAsset("Invalid RIFF format".to_string()));
+                            }
+
+                            let mut reader = CAIReadWrapper {
+                                reader: input_stream,
+                            };
+
+                            // replace/add manifest in memory
+                            let new_contents = inject_c2pa(
+                                &top_level_chunks,
+                                &mut reader,
+                                &[],
+                                Some(new_xmp.as_bytes()),
+                                &self.riff_format,
+                            )?;
+
+                            // save contents
+                            let mut writer = CAIReadWriteWrapper {
+                                reader_writer: output_stream,
+                            };
+                            new_contents
+                                .write(&mut writer)
+                                .map_err(|_e| Error::EmbeddingError)?;
+                            Ok(())
+                        } else {
+                            const MIN_XMP: &str = r#"<?xpacket begin="" id="W5M0MpCehiHzreSzNTczkc9d"?><x:xmpmeta xmlns:x="adobe:ns:meta/" x:xmptk="XMP Core 6.0.0"><rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"><rdf:Description rdf:about="" > </rdf:Description></rdf:RDF> </x:xmpmeta> "#;
```

Comment:
Probably should have this in some common place, I'd imagine we need to do the same thing whenever there isn't already any XMP. Not important as long as this is the only place, but is it?

Reply:
I won't be the only place.  I will figure it out on the next required case.

Reply:
Moved to xmp_utils